File: trunk/sge/SGE/Math/Rot.cpp
#include "Rot.h"
#include "Vec2.h"
#include <cassert>

namespace SGE
{
    
    const Rot Rot::ZERO;
    
    Rot::Rot() :
        angle(0)
    {
    }
    
    Rot::Rot(float _angle) :
        angle(_angle)
    {
    }
    
    Rot::Rot(const Vec2& v)
    {
        angle = Math::atan2(v.y,v.x);
    }
    
    Rot::Rot(const Rot& r) :
        angle(r.angle)
    {
    }
    
    Rot& Rot::operator=(const Rot& r)
    {
        angle = r.angle;
        return *this;
    }
    
    Rot Rot::operator+(const Rot& r) const
    {
        return Rot(angle + r.angle).norm();
    }
    
    Rot Rot::operator-(const Rot& r) const
    {
        return Rot(angle - r.angle).norm();
    }
    
    Rot Rot::operator*(float f) const
    {
        return Rot(angle * f).norm();
    }
    
    Rot Rot::operator/(float f) const
    {
        assert(f != 0);
        return Rot(angle / f).norm();
    }
    
    Vec2 Rot::operator*(const Vec2& v) const
    {
        return Vec2(Rot(v) + *this) * v.length();
    }
    
    Rot Rot::norm() const
    {
        return Rot(Math::fmod(angle,Math::PI));
    }
    
    bool Rot::equals(const Rot& r, float eps) const
    {
        return Math::equal(angle,r.angle,eps);
    }
}

File: pw3/Core/Math/Rot.py
import math

class Rot:
    def __init__(self, a = 0):
        self.a = a
        
    @staticmethod
    def fromDegrees(f):
        return Rot(math.radians(f))
        
    @staticmethod
    def fromRadians(f):
        return Rot(f)
        
    @staticmethod
    def fromVec(v):
        return Rot(math.atan2(v.y,v.x))
    
    def toVec(self, l = 1):
        return Vec2(l * math.cos(self.a), l * math.sin(self.a))
        
    def __repr__(self):
        return "Rot(%.3f)" % self.a
        
    def __eq__(self, r):
        return abs(self.a - r.a) < 1e-4
        
    def __neg__(self):
        return Rot(-self.a)
        
    def __add__(self, r):
        return Rot(self.a + r.a)
        
    def __sub__(self, r):
        return Rot(self.a - r.a)
        
    def __mul__(self, v):
        if isinstance(v,Vec2):
            return Vec2.fromRot(self + v.toRot(),v.length())
        else:
            return Rot(self.a * v)
        
    def __div__(self, f):
        return Rot(self.a / f)
        
    def norm(self):
        return Rot(math.fmod(self.a,math.pi))
        
    def degrees(self):
        return math.degrees(self.a)
        
    def radians(self):
        return self.a

from Vec2 import Vec2
